File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/InformationMessage.java
package org.emulinker.kaillera.controller.v086.protocol;

import com.google.auto.value.AutoValue;
import com.google.common.base.Strings;
import java.nio.ByteBuffer;
import org.emulinker.kaillera.controller.messaging.*;
import org.emulinker.kaillera.relay.KailleraRelay;
import org.emulinker.util.EmuUtil;

@AutoValue
public abstract class InformationMessage extends V086Message {
  public static final byte ID = 0x17;
  private static final String DESC = "Information Message";

  public abstract String source();

  public abstract String message();

  public static AutoValue_InformationMessage create(
      int messageNumber, String source, String message) throws MessageFormatException {
    V086Message.validateMessageNumber(messageNumber, DESC);

    if (Strings.isNullOrEmpty(source)) {
      throw new MessageFormatException("Invalid " + DESC + " format: source.length == 0");
    }

    if (Strings.isNullOrEmpty(message)) {
      throw new MessageFormatException("Invalid " + DESC + " format: message.length == 0");
    }
    return new AutoValue_InformationMessage(messageNumber, ID, DESC, source, message);
  }

  @Override
  public int getBodyLength() {
    return getNumBytes(source()) + getNumBytes(message()) + 2;
  }

  @Override
  public void writeBodyTo(ByteBuffer buffer) {
    EmuUtil.writeString(buffer, source(), 0x00, KailleraRelay.config.charset());
    EmuUtil.writeString(buffer, message(), 0x00, KailleraRelay.config.charset());
  }

  public static InformationMessage parse(int messageNumber, ByteBuffer buffer)
      throws ParseException, MessageFormatException {
    if (buffer.remaining() < 4) throw new ParseException("Failed byte count validation!");

    String source = EmuUtil.readString(buffer, 0x00, KailleraRelay.config.charset());

    if (buffer.remaining() < 2) throw new ParseException("Failed byte count validation!");

    String message = EmuUtil.readString(buffer, 0x00, KailleraRelay.config.charset());

    return InformationMessage.create(messageNumber, source, message);
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/InformationMessage.kt
package org.emulinker.kaillera.controller.v086.protocol

import java.nio.ByteBuffer
import org.emulinker.kaillera.controller.messaging.MessageFormatException
import org.emulinker.kaillera.controller.messaging.ParseException
import org.emulinker.kaillera.controller.v086.protocol.V086Message.Companion.validateMessageNumber
import org.emulinker.kaillera.pico.AppModule
import org.emulinker.util.EmuUtil

data class InformationMessage
    @Throws(MessageFormatException::class)
    constructor(override val messageNumber: Int, val source: String, val message: String) :
    V086Message() {

  override val messageId = ID

  init {
    validateMessageNumber(messageNumber)
    require(source.isNotBlank()) { "source cannot be blank" }
    require(message.isNotBlank()) { "message cannot be blank" }
  }

  override val bodyLength: Int
    get() = getNumBytes(source) + getNumBytes(message) + 2

  public override fun writeBodyTo(buffer: ByteBuffer) {
    EmuUtil.writeString(buffer, source, 0x00, AppModule.charsetDoNotUse)
    EmuUtil.writeString(buffer, message, 0x00, AppModule.charsetDoNotUse)
  }

  companion object {
    const val ID: Byte = 0x17

    @Throws(ParseException::class, MessageFormatException::class)
    fun parse(messageNumber: Int, buffer: ByteBuffer): InformationMessage {
      if (buffer.remaining() < 4) throw ParseException("Failed byte count validation!")
      val source = EmuUtil.readString(buffer, 0x00, AppModule.charsetDoNotUse)
      if (buffer.remaining() < 2) throw ParseException("Failed byte count validation!")
      val message = EmuUtil.readString(buffer, 0x00, AppModule.charsetDoNotUse)
      return InformationMessage(messageNumber, source, message)
    }
  }
}
